Sources/Tracking/Repository/IdentifyRepository.swift

```diff
@@ -0,0 +1,67 @@
+import Foundation
+
+internal protocol IdentifyRepository: AutoMockable {
+    func addOrUpdateCustomer<RequestBody: Encodable>(
+        identifier: String,
+        // sourcery:Type=AnyEncodable
+        // sourcery:TypeCast="AnyEncodable(body)"
+        body: RequestBody,
+        jsonEncoder: JSONEncoder?,
+        onComplete: @escaping (Result<Void, CustomerIOError>) -> Void
+    )
+    func removeCustomer()
+}
+
+internal class CIOIdentifyRepository: IdentifyRepository {
+    private let httpClient: HttpClient
+    private let keyValueStorage: KeyValueStorage
+    private let jsonAdapter: JsonAdapter
+    private let siteId: String
+
+    /// for testing
+    internal init(httpClient: HttpClient, keyValueStorage: KeyValueStorage, jsonAdapter: JsonAdapter, siteId: String) {
+        self.httpClient = httpClient
+        self.keyValueStorage = keyValueStorage
+        self.jsonAdapter = jsonAdapter
+        self.siteId = siteId
+    }
+
+    init(credentials: SdkCredentials, config: SdkConfig) {
+        self.httpClient = CIOHttpClient(credentials: credentials, config: config)
+        self.siteId = credentials.siteId
+        self.keyValueStorage = DITracking.shared.keyValueStorage
+        self.jsonAdapter = DITracking.shared.jsonAdapter
+    }
+
+    func addOrUpdateCustomer<RequestBody: Encodable>(
+        identifier: String,
+        body: RequestBody,
+        jsonEncoder: JSONEncoder?,
+        onComplete: @escaping (Result<Void, CustomerIOError>) -> Void
+    ) {
+        guard let bodyData = jsonAdapter.toJson(body, encoder: jsonEncoder) else {
+            return onComplete(.failure(.http(.noRequestMade(nil))))
+        }
+
+        let httpRequestParameters = HttpRequestParams(endpoint: .identifyCustomer(identifier: identifier), headers: nil,
+                                                      body: bodyData)
+
+        httpClient
+            .request(httpRequestParameters) { [weak self] result in
+                guard let self = self else { return }
+
+                switch result {
+                case .success:
+                    self.keyValueStorage.setString(siteId: self.siteId, value: identifier, forKey: .identifiedProfileId)
+
+                    onComplete(Result.success(()))
+                case .failure(let error):
+                    onComplete(Result.failure(.http(error)))
+                }
+            }
+    }
+
+    func removeCustomer() {
+        keyValueStorage.setString(siteId: siteId, value: nil, forKey: .identifiedProfileId)
```

Comment:
one upshot of storing relative to the site ID is that if we don't clear it out, then there could be old data if someone moves between two site IDs.

Not something to worry about at this stage, 99% of integrations will set & forget a single site ID, but we should have a quick discussion with the squad about what we expect a customer would expect in a scenario where the site ID changes - both in terms of the new state on the device and the existing state

Reply:
Good idea. 

How I handle that in apps that I build is set an expiration on data. If there is data that is considered "too old" in my standards I delete it. When the app starts up I run a quick background task that checks for old data and deletes it. 

Without much thinking about it I would lean towards that. We get to decide what "too old" is. 

The nice thing is that key/value storage is small in size. If it does stick around it will not hurt the device storage much. It's not like we are caching MB of data in a database on the device.